# Bhargav J Shetgaonkar
Implement the n-steps return in the A2C algorithm (5 points)

In [ ]:
!pip install git+https://github.com/BombayBrownBoy09/StableBaseline3_AIPI530

  Cloning https://github.com/BombayBrownBoy09/StableBaseline3_AIPI530 to /tmp/pip-req-build-nnv0rt26
  Running command git clone -q https://github.com/BombayBrownBoy09/StableBaseline3_AIPI530 /tmp/pip-req-build-nnv0rt26
  Created wheel for stable-baselines3: filename=stable_baselines3-1.3.1a0-py3-none-any.whl size=163562 sha256=fbb1803bb6217b4029989d1be9fe15a7d58bfaa787bb0bae936a865cfa67639d
  Stored in directory: /tmp/pip-ephem-wheel-cache-do9exdv3/wheels/0b/08/52/504108cf044fcf5e13bbba7e09163719ef67451817170829eb
Successfully built stable-baselines3


In [ ]:
import gym
import numpy as np
from stable_baselines3 import A2C
from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.evaluation import evaluate_policy

In [ ]:
env = gym.make('CartPole-v1')

# Use a separate environement for evaluation
eval_env = gym.make('CartPole-v1')

N step Advantage

In [ ]:
model = A2C('MlpPolicy', env, verbose=1, tensorboard_log="./a2c_cartpole_tensorboard/", use_n_step_advantage=True)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.


In [ ]:
# Train the agent for 70000 steps
model.learn(total_timesteps=70000)

# Evaluate the trained agent
mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes=100)

print(f"mean_reward using n-step advantage:{mean_reward:.2f} +/- {std_reward:.2f}")

Logging to ./a2c_cartpole_tensorboard/A2C_1
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 28.9     |
|    ep_rew_mean        | 28.9     |
| time/                 |          |
|    fps                | 182      |
|    iterations         | 100      |
|    time_elapsed       | 2        |
|    total_timesteps    | 500      |
| train/                |          |
|    entropy_loss       | -0.655   |
|    explained_variance | 0.0588   |
|    learning_rate      | 0.0007   |
|    n_updates          | 99       |
|    policy_loss        | 0.635    |
|    value_loss         | 17.2     |
------------------------------------
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 34.6     |
|    ep_rew_mean        | 34.6     |
| time/                 |          |
|    fps                | 294      |
|    iterations         | 200      |
|    time_elapsed       | 3        |
|    total_timesteps    | 1000 

/usr/local/lib/python3.7/dist-packages/stable_baselines3/common/evaluation.py:69: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  UserWarning,


mean_reward using n-step advantage:500.00 +/- 0.00


Turning off GAE

In [ ]:
model = A2C('MlpPolicy', env, verbose=1, gae_lambda=0, tensorboard_log="./a2c_cartpole_tensorboard/", use_n_step_advantage=False)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.


In [ ]:
# Train the agent for 70000 steps
model.learn(total_timesteps=70000)

# Evaluate the trained agent
mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes=100)

print(f"mean_reward turning off gae:{mean_reward:.2f} +/- {std_reward:.2f}")

Logging to ./a2c_cartpole_tensorboard/A2C_2
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 24.6     |
|    ep_rew_mean        | 24.6     |
| time/                 |          |
|    fps                | 731      |
|    iterations         | 100      |
|    time_elapsed       | 0        |
|    total_timesteps    | 500      |
| train/                |          |
|    entropy_loss       | -0.692   |
|    explained_variance | -0.0287  |
|    learning_rate      | 0.0007   |
|    n_updates          | 99       |
|    policy_loss        | -0.495   |
|    value_loss         | 12       |
------------------------------------
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 27.2     |
|    ep_rew_mean        | 27.2     |
| time/                 |          |
|    fps                | 760      |
|    iterations         | 200      |
|    time_elapsed       | 1        |
|    total_timesteps    | 1000 

/usr/local/lib/python3.7/dist-packages/stable_baselines3/common/evaluation.py:69: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  UserWarning,


mean_reward turning off gae:9.42 +/- 0.65


Using Default GAE

In [ ]:
# Generalized Advantage Estimator Equivalent to classic advantage when set to 1. 
model = A2C('MlpPolicy', env, verbose=1, tensorboard_log="./a2c_cartpole_tensorboard/", use_n_step_advantage=False)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.


In [ ]:
# Train the agent for 70000 steps
model.learn(total_timesteps=70000)

# Evaluate the trained agent
mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes=100)

print(f"mean_reward default gae:{mean_reward:.2f} +/- {std_reward:.2f}")

Logging to ./a2c_cartpole_tensorboard/A2C_3
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 21       |
|    ep_rew_mean        | 21       |
| time/                 |          |
|    fps                | 725      |
|    iterations         | 100      |
|    time_elapsed       | 0        |
|    total_timesteps    | 500      |
| train/                |          |
|    entropy_loss       | -0.693   |
|    explained_variance | 0.0324   |
|    learning_rate      | 0.0007   |
|    n_updates          | 99       |
|    policy_loss        | 1.8      |
|    value_loss         | 8.45     |
------------------------------------
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 22.5     |
|    ep_rew_mean        | 22.5     |
| time/                 |          |
|    fps                | 759      |
|    iterations         | 200      |
|    time_elapsed       | 1        |
|    total_timesteps    | 1000 

/usr/local/lib/python3.7/dist-packages/stable_baselines3/common/evaluation.py:69: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  UserWarning,


mean_reward default gae:500.00 +/- 0.00


**Results (70k timesteps, 100 episodes) : **

1.   mean_reward using n-step advantage: 500.00 +/- 0.00
2.   mean_reward turning off gae: 9.42 +/- 0.65
3.   mean_reward default gae:500.00 +/- 0.00
